pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,628 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memberlist
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/golang/groupcache/consistenthash"
+	"github.com/hashicorp/memberlist"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha2"
+	crdlister "antrea.io/antrea/pkg/client/listers/crd/v1alpha2"
+	"antrea.io/antrea/pkg/util/k8s"
+)
+
+const (
+	controllerName = "MemberListCluster"
+	// Set resyncPeriod to 0 to disable resyncing.
+	resyncPeriod time.Duration = 0
+	// set default virtual node replicas num of consistent hash, refs https://github.com/golang/groupcache/issues/29
+	defaultVirtualNodeReplicas = 50
+	defaultLeavePeriod         = time.Second
+)
+
+// default Hash Fn is crc32.ChecksumIEEE
+var defaultHashFn func(data []byte) uint32
+
+var typeAssertionErr = errors.New("type assertion error")
+
+type ClusterInterface interface {
+	Run(stopCh <-chan struct{})
+}
+
+type NotifyEventType int
+
+const (
+	NotifyEventTypeNodeJoined NotifyEventType = iota
+	NotifyEventTypeNodeLeave
+	NotifyEventTypeIPPoolChanged
+)
+
+type clusterNodeEventHandler func(objName string, notifyET NotifyEventType)
+
+type hashEventType int
+
+const (
+	ipPoolCreate hashEventType = iota
+	ipPoolUpdate
+	ipPoolDelete
+	nodeLabelUpdate
+)
+
+type clusterHashEvent struct {
+	eventType hashEventType
+	obj       interface{}
+}
+
+type Cluster struct {
+	bindPort int
+	// nodeConfig contains nodeName and ipAddr which would be used
+	NodeConf *config.NodeConfig
+	// Name of local node. Node name must be unique in the cluster.
+	nodeName string
+
+	mlConfig *memberlist.Config
+	mList    *memberlist.Memberlist
+	// conHash hold the consistenthash.Map, when a node join cluster, use method Add() to add some keys to the hash.
+	// when a node leave the cluster, the consistenthash.Map should update
+	conHashMap       map[string]*consistenthash.Map
+	conHashMapRWLock sync.RWMutex
```

Comment:
could we just use `consistentHash` everywhere instead of `conHash`? I think it would improve readability. I see `con` and I keep thinking of "connection".

Reply:
done